src/loaders/gltf.rs

```diff
@@ -0,0 +1,25 @@
+use std::rc::Rc;
+
+use crate::geometry::Mesh;
+use crate::material::Material;
+
+pub fn load_file(filepath: &str) -> Vec<Mesh> {
+    let (document, buffers, _) = gltf::import(filepath).expect("Could not open Gltf file");
+
+    let mut ret: Vec<Mesh> = Vec::new();
+
+    // Load all the materials first, this assumes that the material index
+    // that primitive refers to is loaded in the same order as document.materials()
+    let materials: Vec<_> = document
+        .materials()
+        .map(|material| Rc::new(Material::from_gltf(&material)))
+        .collect();
+
+    for mesh in document.meshes() {
+        for primitive in mesh.primitives() {
+            ret.push(Mesh::from_gltf(&buffers, &primitive, &materials));
+        }
+    }
+
+    ret
```

Comment:
Based on what you said it sounds like refactoring this to use `flat_map` didn't work?

Does replacing the `ret` variable and the nested loop with the following code work?

```rust
document.meshes()
    .flat_map(|mesh| mesh.primitives())
    .map(|prim| Mesh::from_gltf(&buffers, &prim, &materials))
    .collect()
```

If this doesn't work, could you post the compiler error so we can see what's going on?

Reply:
![rust-error](https://user-images.githubusercontent.com/15278132/55672401-b3ece280-5868-11e9-95c0-3d0e630549fb.png)

Reply:
Ah. I see why this is happening. It's because the document doesn't have the meshes stored already. It only parses them from the JSON when we call `meshes`. That means that each `Mesh` only lasts for the scope of that closure. If we return `mesh.primitives()`, we're returning data on the stack that is about to be dropped.

Your nested loop should be fine then. Thanks for trying it out! :smile: